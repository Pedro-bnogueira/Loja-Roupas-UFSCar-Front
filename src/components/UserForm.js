import React, { useState, useEffect } from "react";
import {
    Dialog,
    DialogTitle,
    DialogContent,
    DialogActions,
    TextField,
    Button,
    MenuItem,
    Box,
} from "@mui/material";

const UserForm = ({ open, onClose, mode, formData, onSave }) => {
    const [data, setData] = useState({ ...formData });
    const [errors, setErrors] = useState({});

    // Sincroniza o estado interno com as props formData
    useEffect(() => {
        setData({ ...formData });
        setErrors({});
    }, [formData, open]);

    const handleChange = (e) => {
        const { name, value } = e.target;
        setData({ ...data, [name]: value });
    };

    const validate = () => {
        let tempErrors = {};
        if (!data.name.trim()) tempErrors.name = "O nome é obrigatório.";
        if (!data.email.trim()) tempErrors.email = "O email é obrigatório.";
        if (mode === "create" && !data.password.trim()) {
            tempErrors.password = "A senha é obrigatória no cadastro.";
        }
        setErrors(tempErrors);
        return Object.keys(tempErrors).length === 0;
    };

    const handleSubmit = () => {
        if (!validate()) return;
        onSave(data);
    };

    return (
        <Dialog open={open} onClose={onClose} fullWidth maxWidth="sm">
            <DialogTitle>
                {mode === "create" ? "Adicionar Usuário" : "Editar Usuário"}
            </DialogTitle>
            <DialogContent>
                <Box
                    component="form"
                    sx={{
                        display: "flex",
                        flexDirection: "column",
                        gap: 2,
                        marginTop: 1,
                    }}
                    noValidate
                    autoComplete="off"
                >
                    <TextField
                        label="Nome"
                        name="name"
                        value={data.name}
                        onChange={handleChange}
                        fullWidth
                        error={!!errors.name}
                        helperText={errors.name}
                        required
                    />
                    <TextField
                        label="Email"
                        name="email"
                        type="email"
                        value={data.email}
                        onChange={handleChange}
                        fullWidth
                        error={!!errors.email}
                        helperText={errors.email}
                        required
                    />
                    <TextField
                        label="Senha"
                        name="password"
                        type="password"
                        value={data.password}
                        onChange={handleChange}
                        fullWidth
                        error={!!errors.password}
                        helperText={errors.password}
                        required={mode === "create"}
                    />
                    <TextField
                        label="Nível de Acesso"
                        name="accessLevel"
                        select
                        value={data.accessLevel}
                        onChange={handleChange}
                        fullWidth
                        required
                    >
                        <MenuItem value="user">Usuário</MenuItem>
                        <MenuItem value="admin">Administrador</MenuItem>
                        <MenuItem value="guest">Convidado</MenuItem>
                    </TextField>
                </Box>
            </DialogContent>
            <DialogActions>
                <Button onClick={onClose} color="primary">
                    Cancelar
                </Button>
                <Button
                    onClick={handleSubmit}
                    variant="contained"
                    color="primary"
                >
                    {mode === "create" ? "Adicionar" : "Salvar"}
                </Button>
            </DialogActions>
        </Dialog>
    );
};

export default UserForm;
