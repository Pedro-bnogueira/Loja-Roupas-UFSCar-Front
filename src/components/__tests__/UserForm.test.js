
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import UserForm from '../UserForm'; // Ajuste o path conforme sua estrutura

describe('UserForm Component', () => {
  const mockOnClose = jest.fn();
  const mockOnSave = jest.fn();

  const defaultProps = {
    open: true,
    onClose: mockOnClose,
    onSave: mockOnSave,
    mode: 'create',
    formData: {
      name: '',
      email: '',
      password: '',
      accessLevel: 'user',
    },
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('deve renderizar corretamente no modo "create"', () => {
    render(<UserForm {...defaultProps} />);

    // Título
    expect(screen.getByText(/adicionar usuário/i)).toBeInTheDocument();

    // Botão principal
    expect(screen.getByRole('button', { name: /adicionar/i })).toBeInTheDocument();

    // Campos (usamos regex para ignorar o * que o MUI adiciona)
    expect(screen.getByLabelText(/nome/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/email/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/senha/i)).toBeInTheDocument(); // label no modo create
    expect(screen.getByLabelText(/nível de acesso/i)).toBeInTheDocument();
  });

  it('deve renderizar corretamente no modo "edit"', () => {
    render(
      <UserForm
        {...defaultProps}
        mode="edit"
        formData={{
          name: 'Usuário de Teste',
          email: 'teste@example.com',
          password: '',
          accessLevel: 'admin',
        }}
      />
    );

    // Título
    expect(screen.getByText(/editar usuário/i)).toBeInTheDocument();

    // Botão principal
    expect(screen.getByRole('button', { name: /salvar/i })).toBeInTheDocument();

    expect(screen.getByLabelText(/nome/i)).toHaveValue('Usuário de Teste');
    expect(screen.getByLabelText(/email/i)).toHaveValue('teste@example.com');
    // No modo "edit" deve exibir "Nova Senha"
    expect(screen.getByLabelText(/nova senha/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/administrador/i)).toBeInTheDocument();
  });

  it('deve chamar onClose ao clicar em "Cancelar"', async () => {
    render(<UserForm {...defaultProps} />);
    const cancelButton = screen.getByRole('button', { name: /cancelar/i });

    await userEvent.click(cancelButton);

    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('deve exibir erros de validação se campos obrigatórios estiverem vazios (modo create)', async () => {
    render(<UserForm {...defaultProps} />);

    // Clica em "Adicionar" sem preencher nada
    const addButton = screen.getByRole('button', { name: /adicionar/i });
    await userEvent.click(addButton);

    // Espera pelas mensagens de erro
    expect(await screen.findByText('O nome é obrigatório.')).toBeInTheDocument();
    expect(screen.getByText('O email é obrigatório.')).toBeInTheDocument();
    expect(screen.getByText('A senha é obrigatória no cadastro.')).toBeInTheDocument();

    // onSave não deve ter sido chamado
    expect(mockOnSave).not.toHaveBeenCalled();
  });

  it('deve exibir erros de validação se campos obrigatórios estiverem vazios (modo edit)', async () => {
    render(
      <UserForm
        {...defaultProps}
        mode="edit"
        formData={{
          name: '',
          email: '',
          password: '', // no modo edit não é obrigatório
          accessLevel: 'user',
        }}
      />
    );

    // Clica em "Salvar"
    const saveButton = screen.getByRole('button', { name: /salvar/i });
    await userEvent.click(saveButton);

    // Verifica se aparecem as mensagens de erro para nome e email
    expect(await screen.findByText('O nome é obrigatório.')).toBeInTheDocument();
    expect(screen.getByText('O email é obrigatório.')).toBeInTheDocument();

    // Mas não deve exibir erro de senha obrigatória (no edit)
    expect(screen.queryByText('A senha é obrigatória no cadastro.')).not.toBeInTheDocument();

    // onSave não deve ter sido chamado
    expect(mockOnSave).not.toHaveBeenCalled();
  });

  it('deve chamar onSave com os dados corretos quando o formulário estiver válido (modo create)', async () => {
    render(<UserForm {...defaultProps} />);

    // Preenche os campos (usamos /nome/i em vez de "Nome")
    await userEvent.type(screen.getByLabelText(/nome/i), 'Novo Usuário');
    await userEvent.type(screen.getByLabelText(/email/i), 'novo@teste.com');
    await userEvent.type(screen.getByLabelText(/senha/i), 'senha123');

    // O select
    const accessLevelSelect = screen.getByLabelText(/nível de acesso/i);
    await userEvent.click(accessLevelSelect);
    // Seleciona "Administrador"
    const adminOption = screen.getByRole('option', { name: /administrador/i });
    await userEvent.click(adminOption);

    // Clica em Adicionar
    const addButton = screen.getByRole('button', { name: /adicionar/i });
    await userEvent.click(addButton);

    // onSave deve ter sido chamado com os dados
    expect(mockOnSave).toHaveBeenCalledTimes(1);
    expect(mockOnSave).toHaveBeenCalledWith({
      name: 'Novo Usuário',
      email: 'novo@teste.com',
      password: 'senha123',
      accessLevel: 'admin',
    });
  });

  it('deve permitir salvar sem senha no modo edit', async () => {
    render(
      <UserForm
        {...defaultProps}
        mode="edit"
        formData={{
          name: 'Usuário Existente',
          email: 'exemplo@teste.com',
          password: '', // Sem senha
          accessLevel: 'user',
        }}
      />
    );

    // Clica em "Salvar"
    const saveButton = screen.getByRole('button', { name: /salvar/i });
    await userEvent.click(saveButton);

    // Deve chamar onSave com esses dados
    expect(mockOnSave).toHaveBeenCalledTimes(1);
    expect(mockOnSave).toHaveBeenCalledWith({
      name: 'Usuário Existente',
      email: 'exemplo@teste.com',
      password: '',
      accessLevel: 'user',
    });
  });
});
