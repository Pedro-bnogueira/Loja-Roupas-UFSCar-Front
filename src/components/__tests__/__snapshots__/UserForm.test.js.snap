// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm Component matches snapshot in create mode 1`] = `<DocumentFragment />`;

exports[`UserForm Component matches snapshot in edit mode 1`] = `<DocumentFragment />`;
